Initial commit for plugin internals refactor

Schamper opened the pull request:
Initial commit for the plugin internals refactor. This removes a lot of slow and hard to understand code with (hopefully) faster and easier to understand code. Also fixes some consistency issues I came across while working on this.

There are some things I want to add, but I may do those in a separate PR:
- Re-add support for https://github.com/fox-it/dissect.target/issues/758 (it was lost in between version 2 and 3)
- Add support for "plugin directories", similar to `_os.py`, but then `_plugin.py` or something similar (https://github.com/fox-it/dissect.target/pull/788)
- Add helper functions for getting runtime information from functions, e.g. the class object, the record descriptor objects, etc. This will _finally_ allow us to query plugins by e.g. ones that generate records that have a `path` field, for example. (https://github.com/fox-it/dissect.target/pull/1007)

This should also cover https://github.com/fox-it/dissect.target/pull/759.

Closes https://github.com/fox-it/dissect.target/issues/889.

pyrco commented on tests/test_plugin.py:
```diff
@@ -265,7 +326,7 @@ def test_incompatible_plugin(target_bare: Target) -> None:
 
 
-MOCK_PLUGINS = {
+MOCK_PLUGINS_OLD = {
```
As this is no longer used, it can be removed I think?

pyrco reviewed (changes requested):
I noticed `plugins/general/example.py` and `plugins/os/unix/packagemanager.py` still have a `__findable__` property, which is now no longer needed.

Another thing I noticed is that the `-l` no longer lists the plugins from `_os.py`.

And a third thing I noticed that if I run a namespace plugin, e.g. webserver.access, which is listed when doing `target-query MyTarget -l`, before it printed an error like:

```
Error while parsing sysvol/windows/system32/inetsrv/config/applicationHost.config: ...
```
but now it errors with:
```
target-query: error: argument -f/--function contains invalid plugin(s): webserver.access
```

Schamper commented on dissect/target/plugin.py:
```diff
@@ -262,609 +372,518 @@ def check_compatible(self) -> None:
+def _generate_long_paths() -> dict[str, FunctionDescriptor]:
+    """Generate a dictionary of all long paths to their function descriptors."""
+    paths = {}
+    for value in _get_plugins().get("__functions__", {}).get(None, {}).values():
```
I'm not sure on what the behaviour for that should be, as there are no existing unit tests for it. I assume it's not intended behaviour.

Schamper commented:
> I noticed `plugins/general/example.py` and `plugins/os/unix/packagemanager.py` still have a `__findable__` property, which is now no longer needed.

Fixed package manager but we still don't want the example plugin to findable.
 
> Another thing I noticed is that the `-l` no longer lists the plugins from `_os.py`.

Should be fixed again.

> And a third thing I noticed that if I run a namespace plugin, e.g. webserver.access, which is listed when doing `target-query MyTarget -l`, before it printed an error like:
> 
> ```
> Error while parsing sysvol/windows/system32/inetsrv/config/applicationHost.config: ...
> ```
> 
> but now it errors with:
> 
> ```
> target-query: error: argument -f/--function contains invalid plugin(s): webserver.access
> ```

I had to largely rewrite the `NamespacePlugin` to properly behave without any of its old side effects, but this is now also fixed.

twiggler commented on dissect/target/plugin.py:
```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
+        "__os__": {},
+        # All child plugins
+        # {"<module_path>": PluginDescriptor}
+        "__child__": {},
```
(Nitpick)

Discriminating between child and regular plugins is done for performance it seems? Perhaps somewhat better to  localize the distinction and add memoization to  `child_plugins`.

Schamper replied:
There's no (fast) way to distinct a `ChildTargetPlugin` from a regular plugin once they're in the registry. We only allow basic types in the registry so that it can be easily serialised. Child plugins are special cased and only used in the opening on child targets, so they can't be used in the same places that regular plugins can.

I don't think there's a much better way to mix these.

twiggler commented on dissect/target/plugin.py:
```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
+    },
+    # OS plugin tree
+    # {"<module_part>": {"<module_part>": PluginDescriptor}}
+    "__ostree__": {},
```
This also feels like derived state (i.e. more like a cache) so perhaps it is better to memoize the tree

Schamper replied:
This is indeed a cache, but the important difference between this and memoization is that memoization would need to occur on every first execution, whereas with this we can benefit from a pre-cached index if we deserialise the plugin index.

twiggler commented on dissect/target/plugin.py:
```diff
@@ -264,607 +411,516 @@ def check_compatible(self) -> None:
 
-class OSPlugin(Plugin):
-    """Base class for OS plugins.
+    Additionally, ``register`` sets the following private attributes on the `plugincls`:
```
(Architecture question, somewhat inarticulate)

Should OSPlugins also be plugins? 
Maybe they should be pluggable, but is this registry the right place to handle them?
Because it seems we need to make exceptions for them throughout the functions below.

twiggler replied:
`find_plugin_functions` depends on `__functions__.__os__`.
And there is the `osfilter` in `plugins` and `functions` and `lookup`.

But I don´t have a solution, at this point I cannot even articulate the problem.
Only a nagging feeling.

Let's close for now.

twiggler reviewed (changes requested):
Nice increment.

Left some comments, most are suggestions / ideas, not demands.

QA: did not find regressions but will test again after changes.

Schamper commented on dissect/target/target.py:
```diff
@@ -118,6 +118,46 @@ def __init__(self, path: Union[str, Path] = None):
+        return f"<Target {self.path}>"
+
+    def __getattr__(self, attr: str) -> Union[plugin.Plugin, Any]:
+        """Override of the default ``__getattr__`` so plugins and functions can be called from a ``Target`` object."""
```
I have thought about this too, and this is indeed very annoying. However, I think the experience of being able to access any plugin as an attribute of the `Target` object is exactly what contributes to heavily to the "magic" of `dissect.target`.

An alternative "solution/workaround" I thought about would be a generated `target.pyi` stub file with every possible plugin attribute populated on the `Target` class. This stub file could be CI checked to be up to date with a pytest unit test, similarly to how we e.g. test that all test files are in Git LFS.

I must admit pushing the dependencies to the plugin is interesting from a different perspective. Maybe something to talk about later.

twiggler replied:
> Maybe something to talk about later.

Sure!

> An alternative "solution/workaround" I thought about would be a generated target.pyi stub file with every possible plugin attribute populated on the Target class. This stub file could be CI checked to be up to date with a pytest unit test, similarly to how we e.g. test that all test files are in Git LFS.

Not a bad idea.

Schamper commented on dissect/target/plugin.py:
```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
+PluginDescriptorLookup = TypedDict(
+    "PluginDescriptorLookup",
+    {
+        None: dict[str, PluginDescriptor],
```
I have a slight personal preference for the `__` syntax, to indicate that they are special root keys. But I will change the `None` out.

Schamper commented on dissect/target/plugin.py:
```diff
@@ -264,607 +411,516 @@ def check_compatible(self) -> None:
-    log.debug("Plugin registered: %s.%s", plugincls.__module__, plugincls.__qualname__)
+    yield from (
+        value
+        for key, value in _get_plugins().get("__plugins__", {}).get(index, {}).items()
```
I will try to make this a bit easier. I had different helper functions before, but they were considerable slowdowns. I wanted this function to be as fast as possible, so that's why it ended up being as one list comprehension.

Schamper commented on dissect/target/plugin.py:
```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
+GENERATED = False
+
+
+def export(*args, **kwargs) -> Callable[..., Any]:
```
I'm personally okay with the current state of things, but it is reaching the limits. Let's look at that at some different time.

Schamper commented on dissect/target/plugin.py:
```diff
@@ -71,7 +63,111 @@ class OperatingSystem(StrEnum):
+"""The plugin registry.
+
+Note: It's very important that all values in this dictionary are serializable.
+The plugin registry can be stored in a file and loaded later. Plain Python syntax is used to store the registry.
```
The `TypedDict` is currently only to increase type hinting.

Do you think it should be a concern for this PR? I think it can be handled separately, as I think we won't touch the data structure with that change, just the way it's serialised/deserialised.

Schamper commented on dissect/target/plugin.py:
```diff
@@ -264,607 +411,516 @@ def check_compatible(self) -> None:
     """
+    yield from (
+        value
+        for entry in _get_plugins().get("__functions__", {}).get(index, {}).values()
```
Added some more descriptive names.

Schamper commented on dissect/target/plugin.py:
```diff
@@ -264,607 +411,516 @@ def check_compatible(self) -> None:
 
-    setattr(cls, alias, clone)
 
+def find_plugin_functions(
```
Tried to do what you suggested.

twiggler commented on dissect/target/plugin.py:
```diff
@@ -264,607 +411,516 @@ def check_compatible(self) -> None:
-    If ``osfilter`` is specified, only plugins related to the provided
-    OSPlugin, or plugins with no OS relation are returned.
-    If ``osfilter`` is ``None``, all plugins will be returned.
+    registry = _get_plugins()
```
```
class PluginRegistry:
   plugins: PluginDescriptor[]
   functions: FunctionDescriptor[]
   ostree: dict[str, PluginDescriptor]
   ...

   # Also does serialization and deserialization 
   
   def find_plugin_functions(self, ....)

   def filter_compatible(self, ...)
```

Schamper commented:
> Additionally could `FunctionDescriptor.record` be populated with `__record__` as discussed in #759?

This is not possible since we can't serialise the record descriptor, so it would just be a string. This functionality will be part of an incremental improvement of this PR as discussed in the main PR body:

> - Add helper functions for getting runtime information from functions, e.g. the class object, the record descriptor objects, etc. This will finally allow us to query plugins by e.g. ones that generate records that have a path field, for example.

The goal of this PR is to agree and establish a better base plugin system, and then incrementally improve that.

twiggler reviewed (changes requested):
Code looks good.

Some tests are failing though.

After these have been fixed, I will do another QA pass.

Schamper commented:
And otherwise https://github.com/fox-it/dissect.target/pull/1007 hopefully improves on that 😄
